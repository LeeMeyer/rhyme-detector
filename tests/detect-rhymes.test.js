const detectRyhmes = require("../src/detect-rhymes");
const fs = require("fs");
const path = require("path");

test('Detects rhymes in the man from snowy river', () => {
    let content = fs.readFileSync(path.resolve(__dirname, "the-man-from-snowy-river.txt")).toString();
    let rhymes = detectRyhmes(content);
    expect(rhymes).toMatchSnapshot();
});

test('Detects rhymes in untitled drowning poem', () => {
    let content = fs.readFileSync(path.resolve(__dirname, "untitled.txt")).toString();
    let rhymes = detectRyhmes(content);
    expect(rhymes).toMatchSnapshot();
});


test('Every end rhyme is detected in the man from snowy river', () => {
    let content = fs.readFileSync(path.resolve(__dirname, "the-man-from-snowy-river.txt")).toString();
    let endWords = content.split('\r\n').filter(l => l.match(/(\w+)/g)).map(l => l.match(/(\w+)/g).pop()); 
   
    let rhymes = detectRyhmes(content);
    
    for (let i = 0; i < endWords.length; i+=4)
    {
        let rhymeGroup1 = rhymes.find(r => r.indexOf(endWords[i]) > -1);
        expect(rhymeGroup1).not.toBeNull(); 
        expect(rhymeGroup1).toContain(endWords[i + 2]);

        let rhymeGroup2 = rhymes.find(r => r.indexOf(endWords[i + 1]) > -1);
        expect(rhymeGroup2).not.toBeNull(); 
        expect(rhymeGroup2).toContain(endWords[i + 3]);
    }
})

test("should detect long and short rhyming words", () => {
    let rhymes = detectRyhmes("overnight delight intertwine emboss mine loss shine boss refine incline");
    expect(rhymes.length).toBe(3);

    expect(rhymes[0]).toEqual(expect.arrayContaining([ 'intertwine', 'mine', 'shine', 'incline', 'refine' ]));
    expect(rhymes[0].length).toBe(5);

    expect(rhymes[1]).toEqual(expect.arrayContaining(['emboss', 'loss', 'boss' ]));
    expect(rhymes[1].length).toBe(3);

    expect(rhymes[2]).toEqual(expect.arrayContaining([ 'overnight', 'delight' ]));
    expect(rhymes[2].length).toBe(2);
});

test("should rhyme friendliness with mess", () => {
    let rhymes = detectRyhmes("friendliness mess you're for noobs I guess. and your loser with allow more pests to deploy a mess");
    expect(rhymes.length).toBe(2);
});

test('man from snowy river rhyme detection should run faster than 200ms', async () => {
    let content = fs.readFileSync(path.resolve(__dirname, "the-man-from-snowy-river.txt")).toString();

    let t0 = performance.now();

    detectRyhmes(content);

    var t1 = performance.now();

    expect(t1 - t0).toBeLessThan(200);
});