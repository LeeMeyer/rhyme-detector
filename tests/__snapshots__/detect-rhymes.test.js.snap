// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detects rhymes in the man from snowy river 1`] = `
Array [
  Array [
    "nowhere",
    "there",
    "where",
    "their",
    "hair",
    "air",
  ],
  Array [
    "ride",
    "beside",
    "wide",
    "side",
    "stride",
    "hillside",
    "tried",
  ],
  Array [
    "roam",
    "home",
    "foam",
  ],
  Array [
    "old",
    "hold",
    "cold",
    "household",
  ],
  Array [
    "and",
    "thousand",
  ],
  Array [
    "mountain's",
    "mountains",
    "runs",
  ],
  Array [
    "for",
    "bore",
    "saw",
    "timor",
  ],
  Array [
    "word",
    "mustered",
    "answered",
    "upward",
    "muttered",
    "gathered",
  ],
  Array [
    "passed",
    "fast",
    "past",
  ],
  Array [
    "clump",
    "jump",
  ],
  Array [
    "that",
    "at",
    "wombat",
  ],
  Array [
    "by",
    "try",
    "eye",
    "ply",
    "die",
    "high",
    "sky",
  ],
  Array [
    "had",
    "lad",
    "sad",
    "clad",
  ],
  Array [
    "to",
    "you",
    "view",
    "into",
    "flew",
    "grew",
    "do",
    "too",
    "through",
    "drew",
    "two",
    "who",
    "few",
  ],
  Array [
    "away",
    "fray",
    "they",
    "way",
    "say",
    "may",
    "day",
    "stay",
    "sway",
    "today",
  ],
  Array [
    "no",
    "go",
    "so",
    "snow",
    "throw",
    "overflow",
  ],
  Array [
    "he",
    "fairly",
    "clancy",
    "fiercely",
    "weedy",
    "pony",
    "wiry",
    "we",
    "only",
    "thickly",
    "be",
    "any",
    "snowy",
    "every",
    "many",
    "very",
    "fancy",
    "see",
    "stringy",
    "fiery",
    "boldly",
    "hardy",
    "scarcely",
    "three",
    "frosty",
    "lofty",
    "story",
  ],
  Array [
    "orders",
    "riders",
    "others",
    "timbers",
    "watchers",
  ],
  Array [
    "right",
    "overnight",
    "might",
    "white",
    "sight",
    "bright",
    "slight",
    "midnight",
    "delight",
  ],
  Array [
    "wheel",
    "he'll",
  ],
  Array [
    "riding",
    "wing",
    "ring",
  ],
  Array [
    "keep",
    "deep",
    "steep",
    "sweep",
  ],
  Array [
    "near",
    "cheer",
    "fear",
    "clear",
  ],
  Array [
    "far",
    "are",
  ],
  Array [
    "full",
    "pull",
  ],
  Array [
    "in",
    "even",
  ],
  Array [
    "love",
    "of",
  ],
  Array [
    "horse",
    "racehorse",
  ],
  Array [
    "sudden",
    "one",
    "won",
  ],
  Array [
    "dash",
    "ash",
    "lash",
  ],
  Array [
    "pile",
    "while",
  ],
  Array [
    "man",
    "can",
    "ran",
  ],
  Array [
    "his",
    "is",
    "breezes",
  ],
  Array [
    "then",
    "when",
    "bushmen",
  ],
  Array [
    "but",
    "summit",
  ],
  Array [
    "could",
    "good",
    "stood",
    "would",
  ],
  Array [
    "horsemen",
    "horseman",
  ],
  Array [
    "followed",
    "rode",
  ],
  Array [
    "blood",
    "landed",
    "handed",
    "noted",
    "halted",
    "rugged",
  ],
  Array [
    "black",
    "back",
    "track",
  ],
  Array [
    "thunder",
    "ever",
    "other",
    "were",
    "spur",
    "cur",
  ],
  Array [
    "tread",
    "overhead",
    "said",
    "head",
    "bed",
    "homestead",
    "thoroughbred",
    "bred",
  ],
  Array [
    "reins",
    "plains",
  ],
  Array [
    "never",
    "wherever",
  ],
  Array [
    "echoes",
    "those",
  ],
  Array [
    "him",
    "dim",
  ],
  Array [
    "stand",
    "grand",
    "hand",
  ],
  Array [
    "from",
    "come",
    "bottom",
  ],
  Array [
    "small",
    "all",
  ],
  Array [
    "undersized",
    "prized",
  ],
  Array [
    "horses",
    "horse's",
  ],
  Array [
    "tough",
    "enough",
    "rough",
  ],
  Array [
    "bid",
    "waited",
    "shifted",
    "wanted",
    "unassisted",
    "undaunted",
    "dreaded",
  ],
  Array [
    "will",
    "till",
    "hill",
    "still",
  ],
  Array [
    "gallop",
    "up",
    "cup",
  ],
  Array [
    "well",
    "bridle",
    "battle",
    "terrible",
    "saddle",
    "wistful",
    "final",
    "single",
    "crystal",
    "tell",
  ],
  Array [
    "stop",
    "hop",
  ],
  Array [
    "make",
    "take",
  ],
  Array [
    "breath",
    "death",
  ],
  Array [
    "friend",
    "end",
    "lend",
  ],
  Array [
    "ought",
    "got",
    "brought",
    "trot",
    "hot",
  ],
  Array [
    "let",
    "yet",
    "met",
    "regret",
  ],
  Array [
    "ground",
    "round",
    "found",
    "sound",
    "around",
    "bloodhound",
    "pound",
  ],
  Array [
    "warrant",
    "torrent",
  ],
  Array [
    "he's",
    "gullies",
  ],
  Array [
    "slip",
    "hip",
  ],
  Array [
    "strike",
    "like",
  ],
  Array [
    "swung",
    "among",
  ],
  Array [
    "hills",
    "spills",
  ],
  Array [
    "seen",
    "between",
  ],
  Array [
    "went",
    "sent",
    "descent",
  ],
  Array [
    "brow",
    "now",
  ],
  Array [
    "feet",
    "seat",
  ],
  Array [
    "must",
    "just",
  ],
  Array [
    "pace",
    "face",
    "place",
  ],
  Array [
    "beast",
    "least",
  ],
  Array [
    "mute",
    "pursuit",
  ],
  Array [
    "such",
    "touch",
  ],
  Array [
    "proud",
    "cowed",
  ],
  Array [
    "reveals",
    "heels",
  ],
  Array [
    "heads",
    "beds",
  ],
  Array [
    "alone",
    "own",
  ],
  Array [
    "raise",
    "blaze",
  ],
]
`;

exports[`Detects rhymes in untitled drowning poem 1`] = `
Array [
  Array [
    "seaweed",
    "need",
  ],
  Array [
    "milestone",
    "alone",
    "own",
  ],
  Array [
    "hover",
    "lover",
    "her",
  ],
  Array [
    "sea",
    "me",
    "fatality",
    "reality",
    "body",
  ],
  Array [
    "near",
    "disappear",
  ],
  Array [
    "lung",
    "tongue",
  ],
  Array [
    "flailing",
    "trailing",
  ],
  Array [
    "will",
    "still",
  ],
  Array [
    "pacing",
    "embracing",
  ],
  Array [
    "enters",
    "waters",
  ],
  Array [
    "distort",
    "start",
  ],
  Array [
    "implore",
    "for",
  ],
]
`;
